ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+      | Some (`AntiAffinity group) -> (
+          debug
+            "Spread evenly plan: try to plan for the anti-affinity VM (%s, %s)"
+            (Ref.string_of vm)
+            (Db.VM.get_name_label ~__context ~self:vm) ;
+          try
+            let h =
+              Xapi_vm_helpers.select_host_from_ranked_groups vm host_selector
+                (VMGroupMap.find group grp_ranked_hosts
+                |> IntMap.to_seq
+                |> List.of_seq
+                |> List.fast_sort (fun (a, _) (b, _) -> compare a b)
+                   (* Make sure the ranked list is sorted in terms of vm_count increasingly *)
+                |> List.map snd
+                )
+            in
+            debug "Spread evenly plan: find h: (%s %s)" (Ref.string_of h)
+              (Db.Host.get_name_label ~__context ~self:h) ;
+            let h_size =
+              Int64.sub
+                (List.assoc h hosts_increasing)
+                (List.assoc vm vms_increasing)
+            in
+            debug
+              "Spread evenly plan: choose the host with the least free memory \
+               which can run the vm: %s"
+              (Db.Host.get_name_label ~__context ~self:h) ;
+
+            let hosts_increasing =
+              Binpack.insert bigger_than (h, h_size)
+                (List.remove_assoc h hosts_increasing)
+            in
+            vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+              remaining_vms vm_can_boot_on_host
+              (grp_ranked_hosts |> update_anti_affinity_grp_ranked_hosts group h)
+              ((vm, h) :: mapping)
+          with _ ->
+            debug
+              "Spread evenly plan: no host can run the anti-affinity VM, skip \
+               it, continue to plan for the next one." ;
+            vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+              remaining_vms vm_can_boot_on_host grp_ranked_hosts mapping
+        )
+      | _ ->
+          debug
+            "Spread evenly plan: not an anti-affinity VM, skip it, continue \
+             with the next one." ;
+          vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+            remaining_vms vm_can_boot_on_host grp_ranked_hosts mapping
+    )
+
+(** Try to get a no breach plan for anti-affinity VMs (for each anti-affinity group, there are at least
+    2 hosts having running VMs in the group): 
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, try to select a host for it so that there are at least 2 hosts which
+       has running VMs in the same anti-affinity group. If there are already 2 hosts having running VMs
+       in that group, skip planning for the VM. *)
+let rec vm_anti_affinity_no_breach_plan ~__context hosts_increasing
```

Comment:
I think this function may be implemented equivalently by the `vm_anti_affinity_spread_evenly_plan`/`plan_breadth_first` with pruning the `vms` as:
```
let vms_no_breach groups vms =
  let groups' =
    VMGroupMap.filter_map (fun _group hosts ->
      match HostMap.bindings hosts with
      | [(0, _::_::_)] -> Some 2 (* all VMs run on the hosts to be evacuated *)
      | [(0, _::_); (_, [_])] -> Some 1 (* VMs run on only one host *)
      | _ -> None
    ) groups
  in
  List.fold_left (fun (acc_vms, acc_groups) (vm, vm_size, group) ->
    match VMGroupMap.find_opt group acc_groups with
    | Some n when n > 0 ->
        let acc_vms = (vm, vm_size, group) :: acc_vms in
        let acc_groups = VMGroupMap.update group (Option.map pred) acc_groups in
        (acc_vms, acc_groups)
    | _ ->
        (acc_vms, acc_groups)
  ) ([], groups') vms
  |> fst
```

Reply:
There is one problem which prevents us to predict how many VMs from one group is needed to be planned for a no-breach plan for that group: during the planning, we need to verify if a VM can boot on the planned host, if it can't boot on all hosts, we will need to try the next one from that group.

Reply:
If the smallest one can't, the bigger ones won't.

Reply:
Not failed as not enough memory, but failed by `Xapi_vm_helpers.assert_can_boot_here`.
But just now I checked the code where `Xapi_ha_vm_failover.compute_evacuation_plan` is used, one thing I missed is that, the plan is verified again with `Xapi_vm_helpers.assert_can_boot_here`: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_host.ml#L410 , so one problem I find here is that: for an calculated "spread evenly plan", it might still be not valid one as the planning from binpack has not verified by `Xapi_vm_helpers.assert_can_boot_here` yet.
So I would like to change to verify the plan from binpack with `Xapi_vm_helpers.assert_can_boot_here` before stopping for the next step planning, so that once a "spread evenly plan" or a "no breach" plan is returned, it is guaranteed to be a valid plan.

Reply:
And there is one case I find which prevents us to reuse `vm_anti_affinity_spread_evenly_plan` to make a no-breach plan:
Let's say, in a pool of 3 hosts(host1, host2, host3), for VM anti-affinity group A, we got 3 VMs from that group vm1_grpa, vm2_grpa, vm3_grpa in host1, and vm4_grpa and vm5_grpa in host2, and no VMs from group A in host3. Then if we start to evacuate host1, we can't not only try to plan the smallest VM (among vm1_grpa, vm2_grpa, vm3_grpa) with `vm_anti_affinity_spread_evenly_plan`, as the smallest one might only can boot on host2, with `vm_anti_affinity_spread_evenly_plan`, it will also be planned on host2, while this is not a no-breach plan.

Reply:
Basically, I think the two functions are of quite similar patterns. E.g. 

- iterate on a VM list
- select a host
- update free mems of the selected host
- update placements of the group,
- etc